airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py

```diff
@@ -0,0 +1,260 @@
+
+    @property
+    def cursor_slice(self):
+        return self._cursor_slice
+
+    def __repr__(self):
+        return repr(self._stream_slice)
+
+    def __setitem__(self, key: str, value: Any):
+        raise ValueError("PerPartitionStreamSlice is immutable")
+
+    def __getitem__(self, key: str):
+        return self._stream_slice[key]
+
+    def __len__(self):
+        return len(self._stream_slice)
+
+    def __iter__(self):
+        return iter(self._stream_slice)
+
+    def __contains__(self, item: str):
+        return item in self._stream_slice
+
+    def keys(self):
+        return self._stream_slice.keys()
+
+    def items(self):
+        return self._stream_slice.items()
+
+    def values(self):
+        return self._stream_slice.values()
+
+    def get(self, key: str, default: Any) -> Any:
+        return self._stream_slice.get(key, default)
+
+    def __eq__(self, other):
+        if type(other) == dict:
+            return self._stream_slice == other
+        if type(other) == PerPartitionStreamSlice:
+            # noinspection PyProtectedMember
+            return self._partition == other._partition and self._cursor_slice == other._cursor_slice
+        return False
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+
+class CursorFactory:
+    def __init__(self, create_function):
+        self._create_function = create_function
+
+    def create(self):
+        return self._create_function()
+
+
+class PerPartitionCursor(StreamSlicer):
+    """
+    Given a stream has many partitions, it is important to provide a state per partition.
+
+    Record | Stream Slice | Last Record | DatetimeCursorBased cursor
+    -- | -- | -- | --
+    1 | {"start_time": "2021-01-01","end_time": "2021-01-31","owner_resource": "1"''} | cursor_field: “2021-01-15” | 2021-01-15
+    2 | {"start_time": "2021-02-01","end_time": "2021-02-28","owner_resource": "1"''} | cursor_field: “2021-02-15” | 2021-02-15
+    3 | {"start_time": "2021-01-01","end_time": "2021-01-31","owner_resource": "2"''} | cursor_field: “2021-01-03” | 2021-01-03
+    4 | {"start_time": "2021-02-01","end_time": "2021-02-28","owner_resource": "2"''} | cursor_field: “2021-02-14” | 2021-02-14
+
+    Given the following errors, this can lead to some loss or duplication of records:
+    When | Problem | Affected Record
+    -- | -- | --
+    Between record #1 and #2 | Loss | #3
+    Between record #2 and #3 | Loss | #3, #4
+    Between record #3 and #4 | Duplication | #1, #2
+
+    Therefore, we need to manage state per partition.
+    """
+
+    _NO_STATE = {}
+    _NO_CURSOR_STATE = {}
+    _KEY = 0
+    _VALUE = 1
+
+    def __init__(self, cursor_factory: CursorFactory, partition_router: StreamSlicer):
+        self._cursor_factory = cursor_factory
+        self._partition_router = partition_router
+        self._cursor_per_partition = {}
+
+    def stream_slices(self, sync_mode: SyncMode, stream_state: StreamState) -> Iterable[PerPartitionStreamSlice]:
```

Comment:
```suggestion
    def stream_slices(self, sync_mode: SyncMode, _: StreamState) -> Iterable[PerPartitionStreamSlice]:
```
WDYT about obscuring the argument name to discourage using it?